ISSUE-90: Better handling of synchronous (sequential) requests

gnom7 opened the pull request:
* added support for `extraDuration` feature
* added test for `extraDuration` feature
* `extraDuration` should delay spinner hiding at specified time (in millis), this should help to avoid spinner blinking on sequential requests (when crafting of one request depends on data from previous request, so that between those request may arrive ~5ms gap)

gnom7 commented:
Not sure why one test failed, I run all tests locally and all passes. Should dig into this more.

mpalourdio commented:
Awesome ! I need to give this a shot ! I also need to fix things on my side before, before I'd like we handle these 2 feature requests in a cleaner way, by introducing switchMap, so we can complete/bypass observables. As I am comitting to master sometimes, don't forget to rebase your branch onto from time to time.

FWIW, I need to take decisions on how to mix different options too. Tests should include all mixed cases if we allow mixing, or we need to throw exceptions.

Once again, thanks a lot. Hope I can give this a look next week, that's awesome !

gnom7 commented:
I used same browser and node version as on CI, still no luck to reproduce.

gnom7 commented:
I realized that test failure is not caused by this PR.
How do you intent to use `switchMap`? It is not clear to me.

mpalourdio commented:
Yes, master is not stable. The last beta release has revealed an old and dirty bug. Then I realized the way debounce and delay are handled is bad, and should be gracefully handled before subscription, not at runtime.

It's hard to explain at the moment, working code will be easier to read. I need some time to fix it. I want to get rid of all those 'if' before timers etc.

gnom7 commented:
> introducing switchMap, so we can complete/bypass debounce && delayWhen pipes

updated
> get rid of all those 'if' before timers

updated

tried to use `partition` instead `filter`, but then CI fails for some reason

mpalourdio commented:
Wow, that's definitely awesome ! Thanks a lot !

Would you mind rebasing your branch onto master ? It introduces a failing test-case because of a bug that I've caught after releasing beta yesterday ?

mpalourdio commented:
Yes tests fail locally too, that's expected. You have changed the wrong test. by introducing flush(). Please revert. Existing tests should remain untouched as much as possible.

The failing test is this one : https://github.com/mpalourdio/ng-http-loader/blob/4dc80cde5b37e577f344ee790ae3c94d2723b500/src/test/components/ng-http-loader.component.spec.ts#L500

gnom7 commented:
> You have changed the wrong test

I see, sorry :)
Sadly I'm not able to reproduce test failure locally, so I push changes to remote to trigger CI build

mpalourdio commented:
That's weird. I jun run 'yarn test' or run tests directly from IntelliJ.

We'll cleanup the commit history afterwards,not a big deal.

mpalourdio commented:
I am going to cleanup and push force to your branch if this is OK for you, so I can add commits and have a clean base. Just let me know if this is OK

gnom7 commented:
yes, sure

mpalourdio commented:
Awesome ! But another test is failing now :)

mpalourdio commented:
The test in only failing on chrome, not firefox oO

mpalourdio commented:
The test is fin when ran standalone, but not during the full test suite. Something with dates/ race concurrency I suppose.

gnom7 commented:
I suppose now it is fixed :)

mpalourdio commented:
hey, that's awesome ! You rock :) Merging into master now, so I can add some tests and documentation.

Once again, a big thank you \o/
